### ISSUE TITLE
Provide a trait for getting a type's MIME [discussion]

### ISSUE THREAD

**Issue by blueridanus** [hyperium/mime#79]
**tl;dr:** We should provide a trait that, when implemented, provides a method or a static informing the trait implementor's MIME type. 

---
**some discussion:** HTTP would be pretty limited if it could only meaningfully communicate HTML documents. Beyond just making it possible for browsers to understand multimedia, MIMEs potentially allow all types of transfers to happen in HTTP.

Ideally, this would mean that machines could easily speak to other machines on the web, by allowing them to specify what languages they can use (through _Accept/Content-Type_ headers, for example). That only works, obviously, if they agree on the MIMEs, which is the rationale for having IANA [keep track](https://www.iana.org/assignments/media-types/media-types.xhtml) of them.

Obviously, that is not sufficient for the diversity of application protocols. The convention to deal with this was to prefix **x-** to unofficial MIME types, such as *application/x-mycustomformat*. That convention [was](https://tools.ietf.org/html/rfc6648#appendix-A) recommended, but [has since been deprecated](https://tools.ietf.org/html/rfc6648). Basically, it's counterproductive to the goal of machines speaking to each other, and it pollutes the name space. The new recommendation is to choose a sensible name, assuming it will become a standard or commonly used (because that happens). 

AFAIK, as it is, whenever you want to pass or accept some content through HTTP in Rust in your preferred framework, you probably have to set the headers manually if it's not a trivial format (text/html); or, at least, you have to inform the MIME. That is, if you remembered you should pass meaningful format information in the headers.

We do fantastic with serialization already. All it takes is a `#[derive(Serialize, Deserialize)]` on your struct, and you can pass or accept it through HTTP, for example. But you'd have to remember you need to inform the MIME of this serialized data and manually set the relevant header(s). You also have to guess the MIME if it's not a standard format. 

Even if not too much of a bother, this is problematic, because it likely won't be the same guess another Rust programmer will make on some other application, fragmenting the web and the ecosystem.

The way we do better is by correctly delegating the responsibilities. HTTP frameworks handle setting the headers, and the MIME is informed by a trait implementation on the struct you are trying to pass. That is where _mime_ could come in: providing this trait for other libraries.

**Comment by blueridanus** [comment]
Before a PR, some discussion would be nice. What is the best way to provide this?

**Comment by seanmonstar** [comment]
Thanks for writing up your thoughts!

I'm not sure I understand exactly what you're proposing. I'm not asking for an implementation at all, but do you have tiny code snippets of what you imagine?

I think there could be an issue with assigning a MIME to a struct in the same way you `derive(Serialize)`, because the format the data is output on the wire isn't really relevant to the struct itself. It depends on what serializer is used. If you were to declare a struct as being XML, but then used `serde-json` to serialize it, the declared MIME is incorrect, right?

**Comment by blueridanus** [comment]
Sure! I should've put some code to illustrate. Sorry for another wall of text.

Currently, on a framework such as Rocket, you can have something like:
```rust
#[get("/hello")]
fn hello() -> Json<MyStruct> {
    Json(MyStruct {
        //fields
    })
}
```
On the route "/hello", this will return `MyStruct` as JSON, with the correct MIME on the response's _content-type_ header. Likewise, a route that has `Json<T>` as an input will only catch if it receives the request with the correct _content-type_.

This is correct behaviour, but only possible because it's [specifically handled](https://github.com/SergioBenitez/Rocket/blob/master/contrib/src/json.rs#L110) for JSON. 

To help illustrate, here's some background on my use case: I'm thinking about developing a library on top of Rocket for RESTful services. Besides other features, I wish for it to be usable like this:
```rust
#[rest(get, "/foo")]
fn get_foo() -> Foo {
    Foo {
        //fields
    }
}
```
This returns a Foo in whatever format the client asked for; if there is no handler for the format, the client should be informed with `415 Unsupported Media Type`. If I request with `Accept: application/json`, it returns Foo, JSON serialized. If I request with `Accept: application/xml`, it should return Foo encoded in XML. 

```rust
#[rest(get, "/foo")]
fn get_foo() -> Json<Foo> {
   // ...
}
```
The above would mean the route only returns Foo as JSON, and only does so if the request comes with `Accept: application/json`.

Ideally, there would be no implementation required to support a new format that already has a crate that provides serialization through serde(and there's **a lot**). That means if in our ecosystem, a library for serializing that format exists, we get the ability to correctly handle it in our web applications for free.

For this, all that's needed is that we have the same libraries which inform how to serialize a specific, inform their MIME. For that to be of value, they'd need to inform their MIME in the same way, which is where _mime_ would come in. The bonus is that we should automatically be on the correct foot for a sensible MIME whenever a Rust dev comes up with a new format.

I took this concern to serde initially, but I realized it's not the right place to put that interface. MIMEs are not just for specifying the data encoding of a resource, but also what that resource is through a media type. So, we would like for any crate for a format, such as a library for png, to inform its MIME through that same interface. 

Perhaps what we can expose is a `HasMime` trait, with a required method for getting the MIME?

---
> I think there could be an issue with assigning a MIME to a struct in the same way you derive(Serialize), because the format the data is output on the wire isn't really relevant to the struct itself. It depends on what serializer is used. If you were to declare a struct as being XML, but then used serde-json to serialize it, the declared MIME is incorrect, right?

Hmm. I thought that `Json<MyStruct>` was serde semantics, but it's actually from rocket::Json. Oh.

Well, if we have serde crates do it like (`Format<Content>`), then there's no problem. Format provides `get_mime()`. Anytime a `Format<Content>` is sent through http that implements `Serialize` and `HasMime`, you can have it handled correctly automatically. Plus, thanks to the power of generics, you could have really interesting semantics like:

- `Content` provides a media type (i.e. *application/myformat*)
- `Format<T>` provides a media type (i.e. *application/zip*)
- `impl HasMime for Format<T: HasMime>` is such that `Format<Content>` provides a media type that's `Content`'s media type with `Format`'s media type name as a suffix (i.e. *application/myformat+zip*)

It shouldn't be a problem to get this `Format<T>` semantics into serde libraries, as it's not a breaking change. It's a nice addition too, IMO. Passing serialized information around as Strings and then relying on ourselves to choose the correct deserializer seems unoptimal. Probably not a big issue, but what format a String holds serialized data in should probably be encoded in type information. Obviously, that type should provide a method to get the raw string back.

**Comment by seanmonstar** [comment]
I see what you're looking for now. I agree, it'd be really nice to have such a thing, as it makes it easy to plug in new formats into a web framework.

However, it might be a little difficult to get such a trait in `mime`, or rather, to get random encoders to take a public dependency on `mime`, since it's not quite stable yet. It would probably be best to explore this trait and combination of things in your REST framework, and after lessons learned (and hopefully `mime` has settled on a design), it would be an easier sell.

**Comment by blueridanus** [comment]
Oh. Indeed, that makes sense.

There's also my point about having type-safe representations of encoded data, like `Json<Data>`. Not that it's too relevant to mime, but do you have some input on this?

Also, do you feel comfortable with some initial (feature-gated?) implementation of HasMime here? I could submit a PR.
